### PR TITLE
addmm: Reduce constant time overhead

### PR DESCRIPTION
**Pull request by peterbell10**
Fixes the overhead reported by @ngimel in https://github.com/pytorch/pytorch/pull/40927#issuecomment-657709646

As it turns out, `Tensor.size(n)` has more overhead than `Tensor.sizes()[n]`. Since addmm does a lot of introspection of the input matrix sizes and strides, this added up to a noticeable (~1 us) constant time overhead.

With this change, a 1x1 matmul takes 2.85 us on my machine compared to 2.90 us on pytorch 1.5.

### CODE REVIEW

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 1]
```diff
@@ -350,13 +360,17 @@ Tensor &addmm_cpu_(Tensor& self, const Tensor& mat1, const Tensor& mat2, Scalar
 }
 
 Tensor& mm_cpu_out(Tensor & result, const Tensor & self, const Tensor & mat2) {
-  result.resize_({ self.size(0), mat2.size(1) });
+  TORCH_CHECK(self.dim() == 2, "self must be a matrix");
+  TORCH_CHECK(mat2.dim() == 2, "mat2 must be a matrix");
+  native::resize_(result, {self.sizes()[0], mat2.sizes()[1]});
```
Comment (ngimel) [inline aten/src/ATen/native/LinearAlgebra.cpp:365]: nit: since you are resizing result in addmm_cpu_impl, do you need this resize here?

**Review by ngimel — APPROVED** [review]
Awesome, thanks for fixing!

File: aten/src/ATen/native/LinearAlgebra.cpp (hunk 1 above)
Comment (peterbell10) [reply aten/src/ATen/native/LinearAlgebra.cpp:365]: `result` is also the `self` tensor in `addmm` which must at least broadcast to the correct dimensions.

**Comment by ezyang** [comment]
> As it turns out, Tensor.size(n) has more overhead than Tensor.sizes()[n]

Ugggh. Is this because size(n) is doing checked index access and [n] is doing unchecked indexing? Another reason we need to devirtualize size so that compiler can optimize away bounds test if you check dim earlier...

**Comment by ngimel** [comment]
In addition to checked access, Tensor.size(n) also maybe_wraps n, for additional overhead.

**Comment by peterbell10** [comment]
`size(n)` does dim wrapping and bounds checking but is also not inlined and does runtime dispatch for each call.

**Pull request closed without merging**